Vacation PR: multi-paged, styled, searchable docs with wrapper tooling and epubs for NixOS and Nixpkgs

grahamc opened the pull request:
TLDR:

 - New nixpkgs manual: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixpkgs/doc/nixpkgs-manual/html/index.html
 - Nixpkgs manual, epub: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixpkgs/doc/nixpkgs-manual/nixpkgs-manual.epub
 - New NixOS manual: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixos/doc/nixos-manual/html/index.html
 - NixOS manual, epub: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixos/doc/nixos-manual/nixos-manual.epub

###### Motivation for this change

 - The generated doc pages are huge and don't get indexed. 
 - Ctrl-F is a decent search tool, but a proper search engine with stemming is better.
 - Smaller pages are easier to navigate.
 - The output of `nix-build ` was boring.

###### Things done

1. Introduce a `docbook-index` tool for indexing DocBook against the generated HTML.
2. Split up the outputted documentation in to many smaller pages.
3. Added a whole new layout and theme to the docs which is consistent across NixOS and Nixpkgs.
4. Added a local, in-browser search engine (Elasticlunr, what Rust uses.)
5. Way reduced the complexity of generating the Nixpkgs and NixOS docs.
6. Made the Nixpkgs and NixOS docs build the same way.
7. Created `nix-doc-tools`, a metapackage used for building our documentation.
8. Updated the `how to write docs for nixos` docs to account for the `nix-doc-tools` tooling updates. 

BTW: nixpkgs and nixos docs are built the same way, so the docs apply to both.

Major thanks to @samueldr for the work on this.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

NOT DONE:
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
Need to determine the impact of building these docs on the minimal ISO.

I'm going on vacation, so I'll be back on the 16th or so to see what y'all think.

_btw, samueldr can make fixes to this branch for style/js things :)_

grahamc committed:
docs: initial search hacks

grahamc committed:
I think the parameters now chunk with full tocs?

grahamc committed:
params: use UTF-8 content type

grahamc committed:
params: go back to the standard toc-context

grahamc committed:
docs: switch to toc.xml for a page-per-toc-enttry

grahamc committed:
Generate search index each build

grahamc committed:
style: move toc on left

grahamc committed:
Revert "docs: switch to toc.xml for a page-per-toc-enttry"

Not quite right for now.

This reverts commit f3249652c5ff9ef55f62845172b18fd4411ab945.

grahamc committed:
Revert "style: move toc on left"

just busted!

This reverts commit 76932b6f3d023537112fc7d9e7754567a61d12c8.

grahamc committed:
docs makefile: ensure every page has at least one(lol) toc

grahamc committed:
Make each chunk get a toc

grahamc committed:
Toc: add 'current-page' class to the currently active element

grahamc committed:
If the current node in the DOM contains the context in which the TOC is being written, mark it as the current section

grahamc committed:
Swap from dd/dt to ul/li

grahamc committed:
apply current-section to current-page too

grahamc committed:
switch to chunkfast since we use xsltproc

grahamc committed:
Stuff from samueldr: hiding irrelevant things (but lol breaks on the root page)

grahamc committed:
docbook-index: init at 0.1.0

grahamc committed:
Switch to docbook-index package

grahamc committed:
Move params in to the store, was helpful for debugging, not needed anymore

grahamc committed:
delete index from the doc-support

grahamc committed:
docbook-index: 0.1.0 -> 0.1.1

grahamc committed:
Rip off mdBook's teaser code

grahamc committed:
docbook-index: 0.1.1 -> 0.1.2

grahamc committed:
Format doc/manual.xml

samueldr committed:
doc: refactors search.js as strict and self-contained

grahamc committed:
fixups

grahamc committed:
docbook-index: 0.1.2 -> 0.1.3

grahamc committed:
nixpkgs docs: update docbook-index integration

samueldr committed:
docs: Handle style.css in doc-support

samueldr committed:
docs: Manage styles using less

samueldr committed:
docs: Completely restyle docs

samueldr committed:
docs: Make search togglable

samueldr committed:
docs: Enhance navigation

samueldr committed:
docs: Add class when JS is active

samueldr committed:
docs: Fix viewport in JS

samueldr committed:
docs: Add a global menu

samueldr committed:
docs: Clean up behind the developer

A developer that is hacking on the documentation stuff may inadvertantly
have a `out` directory, or `result` symlink that breaks the build.

samueldr committed:
docs: The main page is index.html

grahamc committed:
doc-support: make more generic by passing the locationsXml and functionDocs in to a buildEnv separately

grahamc committed:
doc-support: initial move (wip)

grahamc committed:
nix-doc-tools: accept tools over the arg instead of { nixpkgs } (wip)

grahamc committed:
nix-doc-tools: fixup (wip)

grahamc committed:
nix-doc-tools: dupe xmlformat (wip)

grahamc committed:
nixpkgs docs: use nix-doc-tools

grahamc committed:
nixos docs:  hacky first support for this new shared tool

grahamc committed:
nixos doc: generated refs to doc-support

grahamc committed:
Expand title and body fields

grahamc committed:
nixpkgs docs: switch to nix-doc-tools builder

grahamc committed:
Build the nixpkgs docs with the new builder, pt2

grahamc committed:
create a docs-generate to do all the dynamic generation steps

grahamc committed:
nix-doc-tools: add hints about docs-debug

grahamc committed:
docbook-index: 0.1.3 -> 0.1.4

grahamc committed:
nix-doc-tools: Make preBuild optional

grahamc committed:
nix-doc-tools: add docs-format

grahamc committed:
nixos docs: use nix-doc-tools

grahamc committed:
fixup 70def03ea0491fa08f18c997b7fa9326b7fd07d5

grahamc committed:
nixos docs: Document using the new docs tooling

grahamc committed:
Fixup: nixos docs are called nixos, not nixpkgs

grahamc committed:
nixpksg docs: drop old bits of the makefile

jonringer commented:
manpages no longer exists on manual attr set:
https://github.com/NixOS/nixpkgs/blob/be8bd43041c834dcb47a11eb43d14cf225604ee2/nixos/modules/misc/documentation.nix#L194

I only see manual, manualHTML, manualHTMLShell
https://github.com/NixOS/nixpkgs/blob/be8bd43041c834dcb47a11eb43d14cf225604ee2/nixos/doc/manual/module.nix#L41

veprbl commented:
Will this break all of the current URLs that refer to the anchors on the single page?

samueldr commented:
> Will this break all of the current URLs that refer to the anchors on the single page?

Yes, as it is. I *think* we could work out a solution, literally thinking of it as I type.

 1. Prepare an index of anchors (they should still be named the same)
 1. On index.html, at load, cross-reference anchor with index, redirect.

The main issue there could be is history hijacking. It *may* be possible to do something tricky with html5 history, and window.reload, to overwrite the one entry.

###### Tasks

 * [ ] Implement JS for redirection on index
 * [ ] Implement anchors index

davidak commented:
Great work!

Can we have the search box in the header like this?

![Screenshot from 2019-09-04 18-38-28](https://user-images.githubusercontent.com/91113/64274321-6ec9fb00-cf43-11e9-9820-49228d2c9296.png)

samueldr commented:
It might be possible, but it will cause issues with narrower layouts, especially with sections having longer titles.

There is another issue with the UX where the current implementation presents a clear separation of the search *mode* and the documentation consultation *mode*.

And additionally, while it looks like there is only a search box, there are elements in the search overlay that are currently unlikely to be seen. Among those is the loading state of the index. If we want to add more options to the search engine we would be furthermore constrained in space.

So, the short of it is that horizontal space is a premium, and it can't easily be spared for the search box. I did initially try it in the header, but it didn't feel right.

grahamc commented:
_materializes from vacation for a few minutes_ hmm not a lot of feedback. I hope this can get in for 19.09! :) _disappears back to vacation_

toonn commented:
Some feedback from someone who uses a narrow (for today's standards) browser. The index takes up a lot of horizontal space when you're not using it. [The python docs handle this really well](https://docs.python.org/3/library/re.html) by allowing you to fold the index.

Another thing that has always kinda annoyed me, the toc is super long (multiple accelerated scrolls long), and beginners have no need for it. The main page imo should either be the first section or a short explanation of the important bits of the manual, describing what's in the *nixpkgs* manual and that other manuals exist, e.g., nix and nixos manuals respectively and why they might want to read those instead (similar treatment for the other manuals).

samueldr commented:
> Some feedback from someone who uses a narrow (for today's standards) browser. The index takes up a lot of horizontal space when you're not using it. The python docs handle this really well by allowing you to fold the index.

I guess you're not going narrow enough :)

![image](https://user-images.githubusercontent.com/132835/64572814-e37fa980-d336-11e9-804d-ab486d0b369c.png)

I do use a narrow column, often, and have chosen what felt like an appropriate breakpoint to put the menu away. Though, I do agree that (maybe as a later step) a control could be added to toggle the menu at any levels.


###### Tasks

 * [ ] Index collapse/open control

jtojnar reviewed (commented):
Big :+1: for fixing https://github.com/NixOS/nixpkgs/issues/68199 even before I opened it.

The tools look great, the only pain point is that `docs-build` is somewhat slower than `make out/html/index.html`. Maybe it would make sense to split the HTML and EPUB builds. I dream of fast rebuilds with live reloading the HTML page but I doubt there are free XSLT processors that support incremental builds.

toonn commented:
> I guess you're not going narrow enough : )
@samueldr apparently it's about a cm's worth of difference. But when you go narrower you lose the index completely? The nice thing about the python docs is that you can *choose* whether you want the index visible or not and that it's never inaccessible.

samueldr committed:
nix-doc-tools: Be less noisy with interactive usage

Though, this is only a *default*, prefixing with DEBUG=1 will re-add the
expected noise.

samueldr committed:
nix-doc-tools: docs-build shows a useful error message

samueldr committed:
nix-doc-tools: adds a shell hook for user-friendlyness.

samueldr committed:
nix-doc-tools: search.js -> docs.js

It initially held the code for the search engine only. It does not
anymore.

samueldr committed:
nix-doc-tools: Handle meta viewport in xsl

samueldr committed:
nix-doc-tools: docbook_xsl_ns → docbook-xsl-ns

samueldr committed:
nix-doc-tools: prefer using querySelector

samueldr committed:
nix-doc-tools: adds xhtml namespace to xsl

samueldr commented on pkgs/tools/misc/nix-doc-tools/default.nix:
```diff
@@ -27,15 +27,31 @@ let
+    trap finish EXIT
+
+    input="${root-file-name}"
+    output="$1"
```
As long as graham is fine with the change, a message has been added.

 * https://github.com/NixOS/nixpkgs/commit/d52120bc05ceee022a903432bd8a4e2964fb2a2e

samueldr commented:
I have marked a bunch of @jtojnar's suggestions as resolved, they were trivial changes or changes I approve, in this WIP branch.

 * https://github.com/grahamc/nixpkgs/compare/samuel-docs/searchable-2...samueldr:feature/docs/searchable-3

There are additional changes in there that I would prefer graham would take a look first :)

samueldr committed:
nix-doc-tools: (JS) redirects old anchors to new locations

It is lacking:

 * Indexing the anchors/location pairs

samueldr commented:
Answering @toonn's question:

> But when you go narrower you lose the index completely?

Not at all! There are three ways the narrow view allows the user to navigate:

  * The navigation icons at the top, to move to siblings or parent nodes.
  * The local TOC gets expanded at the top of the content, for easy visualization of the parent nodes, and the document's nodes.
  * The full TOC is present at the end of the document.

In fact, the narrow view shows more to the user, compared to the full view. Though, in reality the additional navigation at the end of the page is not as trivial to use.

> The nice thing about the python docs is that you can choose whether you want the index visible or not and that it's never inaccessible.

And you're right, it's nice to have the option to toggle it :).

samueldr committed:
docs: Adds toggle widget for the TOC

samueldr committed:
docs: The TOC is togglable and hidden by default on narrow widths

samueldr commented:
@toonn → https://stuff.samueldr.com/nixpkgs/manual/

 * :heavy_check_mark: That version has the togglable TOC.
 * :heavy_check_mark: That version has redirection.

To test the redirection:

 * https://stuff.samueldr.com/nixpkgs/manual/index.html#sec-emacs
 * https://stuff.samueldr.com/nixpkgs/manual/#sec-emacs

**Both** index.html and index.html-less  versions should redirect to the right location. That is because both index.html and index.html-less versions of the page are served by the nixos.org server.

I made sure to host the page at a location mimicking the URL structure from nixos.org.

toonn commented:
@samueldr, very nice. All I'm missing is a friendlier "main" page than a detailed TOC like I mentioned earlier and possibly having toggles in the sidebar to open up items and dig down to sub sections. But this is already a major improvement imo. 👍

samueldr commented:
> All I'm missing is a friendlier "main" page

This is out of scope of the PR; contents (except contribution to the documentation) is not part of this PR, considering this reviews the way docs function, both building and using documentation. Though I agree that this is a desirable improvement, to have something more on the main page.

> [...] having toggles in the sidebar to open up items and dig down to sub sections.

It was tried, but does not work well. Every level of the documentation has a page it links to. This meant that a control only used to toggle the open state of a section in the TOC would have to be implemented. Instead, I deemed it was more user-friendly to simply have one action for the options in the menu "follow and expand". With a local file, the navigation **is** instant, meaning there is no drawback. With hosted files (e.g. nixos.org) the navigation is near instant, due to the major decrease in page size. The cost in maintaining an additional non-native feature to the documentation is imho not worth the ability the collapse/expand TOC sections at will. That is especially true considering that following a link loses the expanded/collapsed state.

toonn commented:
Then perhaps a static extra level to the TOC? The current sections are too broad imo, making the sidebar less useful.

grahamc committed:
docbook-index: 0.1.4 -> 0.1.5

grahamc committed:
nix-doc-tools: build man pages

grahamc committed:
nixos docs: split `man` pages in to a separate directory

They were always a separate build anyway, but this way it is easier to
tell.

samueldr committed:
nix-doc-tools: Produce an anchors index

What's the deal with the `cd`??

The filename used in the index are relative to cwd/the given path. This
means that to produce the expected filename associations, we need to be
in the right directory when producing the index.

Alternatively, the indexing tool could have been made to strip the
leading path from the filenames.

jtojnar commented on doc/Makefile:
```diff
@@ -1,90 +1,7 @@
+all: ${MD_TARGETS}
 
 %.section.xml: %.section.md
 	pandoc $^ -w docbook+smart \
```
According to the manual pandoc does not support smart extension for docbook.

https://pandoc.org/MANUAL#typography

grahamc committed:
nix-doc-tools: add timing information to each xsltproc step

grahamc committed:
nix-doc-tools Document what the make.toc override does differently

grahamc committed:
nix-doc-tools: Disable universal toc override since it is too slow.

grahamc committed:
nix-doc-tools: allow passing in olinkdb entries

Allows for cross-referencing documentation. The way
this is built, nixpkgs can depend on nixos, nixos can
depend on nixpkgs, without an infinite loop.

grahamc committed:
nixos docs: Use olinkdbs for the manual in manpages and manual

grahamc committed:
nixos: fixup manualHTMLIndex

grahamc commented:
- [ ] `  <xsl:template name="make.toc">` is way too slow right now. Disabled for development, but needs to be re-enabled so the ToCs work properly.

Mic92 commented:
@grahamc is it known why xslt is too slow. Could be there a switch that only generates the expensive variant for nixos.org and not for local use?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/documentation-presentation-navigation/6401/3

Mic92 commented:
Should we close this PR?

jonringer commented:
It does have a large number of conflicts. And not sure what direction the docs are going after the recent rfc and marketing team changes

cc @garbas

jonringer commented:
@grahamc  your time and effort is still appreciated. Thank you for your hard work :)

grahamc commented:
:) it was useful to me even if for no other reason than to convince me docbook isn’t worth the fight :). 

On Thu, May 28, 2020, at 6:31 PM, Jon wrote:
> 

> @grahamc <https://github.com/grahamc> your time and effort is still appreciated. Thank you for your hard work :)


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NixOS/nixpkgs/pull/68020#issuecomment-635642825>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAASXLFH6WIZDEZAX2MXHJDRT3Q5FANCNFSM4ITJCMNQ>.
